### PR TITLE
Use heap in priority queue implementation

### PR DESCRIPTION
**Pull request by dalg24**
~NOTE: If `swap()` turn out to be a hot spot, consider replacing with simple copy assignment and directly copy inserted element at the end when position is known.~

### CODE REVIEW

**Comment by dalg24** [comment]
> I wouldn't assume the priority queue implementation in the test.

I am not assuming the implementation here.  If you push elements one by one, which I do, heap is fully determined.

File: packages/Search/src/details/DTK_DetailsPriorityQueue.hpp [hunk 1]
```diff
@@ -22,6 +23,16 @@ namespace DataTransferKit
...
+// FIXME probably doesnt belong here
+// also not sure if I actually cannot use std::move() with CUDA
+template <typename T>
+KOKKOS_INLINE_FUNCTION void swap( T &a, T &b )
+{
+    T c( Kokkos::Impl::move( a ) );
+    a = Kokkos::Impl::move( b );
+    b = Kokkos::Impl::move( c );
+}
+
```
Comment (dalg24) [inline packages/Search/src/details/DTK_DetailsPriorityQueue.hpp]: c.f. kokkos/kokkos#1622

**Comment by dalg24** [comment]
@hydroo would you give a shot to de7dc2a?  I got rid of all the calls to `swap`, traded them for copy assignments and delayed copy of the element that was swapped with others to the end when we know its final new position.

**Comment by hydroo** [comment]
Summitdev, GCC 5.4.0, 100M values, 1M queries

What      | master | #392  | 04ad43b | de7dc2a
--------- | ------ | ----- | ------- | -------
Serial    | 1      |  .941 |  .927   |  .887
OpenMP 1  | 1      |  .949 |  .927   |  .888
OpenMP 2  | 1      |  .951 |  .929   |  .889
OpenMP 4  | 1      |  .947 |  .929   |  .892
OpenMP 8  | 1      |  .950 |  .939   |  .899
OpenMP 16 | 1      |  .963 |  .952   |  .907
CUDA      | 1      | 1.163 | 1.150   | 1.117

[Raw data](https://github.com/ORNL-CEES/DataTransferKit/files/2005710/bvh_1.xlsx) (On the bottom)

**Comment by aprokop** [comment]
Further optimizations and general algorithm improvement may be possible by using the fact that we are not inserting random values. Lets look at the actual piece of code that uses `push` and `pop`.
```c++
 while ( !queue.empty() && count < k )
    {
        // get the node that is on top of the priority list (i.e. is the
        // closest to the query point)
        Node const *node = queue.top().first;
        double const node_distance = queue.top().second;
        // NOTE: it would be nice to be able to do something like
        // tie( node, node_distance = queue.top();
        queue.pop();
        if ( TreeTraversal<DeviceType>::isLeaf( bvh, node ) )
        {
            insert( TreeTraversal<DeviceType>::getIndex( bvh, node ),
                    node_distance );
            count++;
        }
        else
        {
            // insert children of the node in the priority list
            for ( Node const *child :
                  {node->children.first, node->children.second} )
            {
                double const child_distance = distance( child );
                queue.push( child, child_distance );
            }
        }
}
```
We see that in many situation `pop` is immediately followed by two `push`es of children. Now consider a situation when a distance to the child (`child_distance`) is larger than the distance to parent (`node_distance`). In this case, one can replace `pop` and `push` with a single `pop_push` which replaces the top element with the child.

In general, we have a `pop_push2` kind of thing, and there potentially may be multiple optimizations depending on how distance to the parent compares to distances of children. But even the above simple approach should provide speedup.

**Comment by dalg24** [comment]
Didn't mean to close the PR...  @aprokop This may be beneficial on the CPU but we see that all our attempts to optimize have led to execution divergence which kills us on the GPU.  You are welcome to piggyback on this PR if you want to experiment some changes.

**Comment by sslattery** [comment]
Based on this we see some significant algorithmic differences between the hardware types - which makes sense. Are we planning on specializing some of these operations on the `Device` type so we can choose the best algorithm for the hardware?

**Comment by aprokop** [comment]
@dalg24 You would probably get execution divergence in priority queue anyway.

**Comment by aprokop** [comment]
Another question about the whole test: the driver populates the source with random points, and then searches for those points. Isn't that a problem on the device that randomness means that queries are not physically local, and thus we almost immediately hit divergence on GPUs? The ideal case for BVH is probably sorting the queries by location before searching. Maybe the driver needs to be updated for that use case, otherwise we never see performance improvements.

**Comment by sslattery** [comment]
What about an option in the driver to spatially sort or leave as random? I would actually like to see what the difference is.

**Comment by aprokop** [comment]
@hydroo Need more comments in `pop_push`, describing the relationship of putting new element at the end of the heap and its relationship with `push`.

**Comment by dalg24** [comment]
Ready for review

File: packages/Search/src/details/DTK_DetailsPriorityQueue.hpp [hunk 2]
```diff
@@ -36,55 +36,110 @@ template <typename T, typename Compare = Less<T>>
...
+            // compare the new root with its children
+            // if they are in the correct order, stop
+            // if not, swap the element with one of its children and continue
+            IndexType const left_child = 2 * pos + 1;
+            IndexType const right_child = 2 * pos + 2;
+            IndexType next_pos = _size - 1;
+            for ( IndexType child : {left_child, right_child} )
+                if ( child < _size - 1 &&
+                     _compare( _heap[next_pos], _heap[child] ) )
+                    next_pos = child;
```
Comment (aprokop) [inline packages/Search/src/details/DTK_DetailsPriorityQueue.hpp]: I feel like there is some potential for optimization here. If we check right child first, and `_compare` is "greater than" operation, do we still need to check `left_child`? I.e., would something like this work?
```c++
if (right_child < _size-1 && _compare(_heap[next_pos], _heap[right_child])
  next_pos = right_child;
else if (left_child < _size-1 && _compare(_heap[next_pos], _heap[left_child])
  next_pos = left_child;
```

File: packages/Search/src/details/DTK_DetailsTreeTraversal.hpp [hunk 3]
```diff
@@ -177,22 +178,25 @@ nearestQuery( BoundingVolumeHierarchy<DeviceType> const &bvh,
...
-                  {node->children.first, node->children.second} )
-            {
-                double const child_distance = distance( child );
-                queue.push( child, child_distance );
-            }
+
+            auto const left_child = node->children.first;
+            auto const right_child = node->children.second;
+            queue.pop_push( left_child, distance( left_child ) );
+            queue.push( right_child, distance( right_child ) );
```
Comment (aprokop) [inline packages/Search/src/details/DTK_DetailsTreeTraversal.hpp]: Can some smart person figure out whether it's generally a better idea to first push left or right child here? Or does it matter?

File: packages/Search/test/tstDetailsTreeTraversal.cpp [hunk 4]
```diff
@@ -58,3 +60,266 @@ TEUCHOS_UNIT_TEST( LinearBVH, priority_queue )
...
+    // unlikely that an error will be raised (this would happen if pop() is
+    // called on an empty queue or push(x) on a queue that is already at maximum
+    // capacity)
+    for ( int i = 0; i < 64; ++i )
+    {
+        queue.push( uniform_distribution( generator ) );
+        check_heap( queue, success, out );
+    }
+
+    // chose randomly whether to call pop(), push(x), or pop_push(x) and check
```
Comment (aprokop) [inline packages/Search/test/tstDetailsTreeTraversal.cpp]: chose -> choose

**Comment by aprokop** [comment]
I would still like to see the update to the driver that searches for spatially local points to demonstrate the performance on GPUs. I understand that it can be done in a separate PR.

File: cmake/Dependencies.cmake [hunk 5]
```diff
@@ -1,9 +1,9 @@
 TRIBITS_PACKAGE_DEFINE_DEPENDENCIES(
   SUBPACKAGES_DIRS_CLASSIFICATIONS_OPTREQS
     Utils                 packages/Utils                      ST  REQUIRED
-    Interface             packages/Interface                  ST  REQUIRED
+    Interface             packages/Interface                  ST  OPTIONAL
```
[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/Dependencies.cmake]: I will drop that commit.

File: packages/Search/src/details/DTK_DetailsPriorityQueue.hpp [hunk 6]
```diff
@@ -36,55 +36,110 @@ template <typename T, typename Compare = Less<T>>
...
+        IndexType pos = 0;
+
+        // perform down-heap operation
+        while ( true )
+        {
+            // compare the new root with its children
+            // if they are in the correct order, stop
+            // if not, swap the element with one of its children and continue
+            IndexType const left_child = 2 * pos + 1;
+            IndexType const right_child = 2 * pos + 2;
```
Comment (sslattery) [inline packages/Search/src/details/DTK_DetailsPriorityQueue.hpp]: `IndexType const right_child = left_child + 1`?

**Comment by dalg24** [comment]
Here is what I measure on the dgx station (that has Volta GPUs) with `--values=100000000` and  `--queries=10000000`

timing\version | current master (d0cdca0) | current version of this PR rebased on master (8edc76a)
--- | --- | ---
construction | 21.609 | 21.4668
kNN | 7.96706 | 3.49588
radius | 3.35762 | 3.39032

**Comment by dalg24** [comment]
On Kepler with `--values=10000000 --queries=100000` (i.e. 10X less values and 10X less queries because we ran out of memory)

timing\version | current master (d0cdca0) | current version of this PR rebased on master (8edc76a)
--- | --- | ---
construction | 1.17461 | 1.17597
kNN | 3.2622 | 1.6589
radius | 2.11633 | 2.11624

**Comment by aprokop** [comment]
@dalg24 In your experiments, did you run it multiple times? If yes, what's the variation in times?

**Comment by hydroo** [comment]
@dalg24 Any idea why it is so much different from my measurements.
Your measurements are on GPU, right?
Could you do CPU measurements, too?

**Comment by dalg24** [comment]
> @dalg24 In your experiments, did you run it multiple times? If yes, what's the variation in times?

couple times, variation was of the order of what you see in the table for "construction' and "radius"

**Comment by dalg24** [comment]
> @dalg24 Any idea why it is so much different from my measurements.

No

> Your measurements are on GPU, right?

Yes

> Could you do CPU measurements, too?

OpenMP timing are not as stable when I work in containers but I guess I could try.

**Comment by hydroo** [comment]
Summitdev, Pascal, 10M values, 100k Queries

**timing/version** | d0cdca0 KNN | 8edc76a KNN | normalized
------------------ | ----------- | ----------- | ----------
Serial             | 1.277       | .948        | .74
OpenMP 1           | 1.276       | .947        | .74
OpenMP 2           | .665        | .489        | .74
OpenMP 4           | .324        | .242        | .75
OpenMP 8           | .164        | .124        | .76
OpenMP 16          | .087        | .066        | .76
CUDA               | .119        | .047        | .39

Can confirm @dalg24 's results.
I had `-g` in the build, but it shouldn't influence CUDA performance.
Who knows.

Congratz on the great speedup!

File: packages/Search/src/details/DTK_DetailsTreeTraversal.hpp (hunk 3 above)
Comment (hydroo) [reply packages/Search/src/details/DTK_DetailsTreeTraversal.hpp]: I'll take the bait.
I measured the exact same for CUDA, and 0-3% slower on CPU when pop_pushing the right child first.

Comment (aprokop) [reply packages/Search/src/details/DTK_DetailsTreeTraversal.hpp]: @hydroo Thanks. It's good to know that it makes no difference.

**Pull request merged**